pkg/tracing/jaeger/jaeger.go

```diff
@@ -33,6 +43,94 @@ func (t *Tracer) GetTraceIDFromSpanContext(ctx opentracing.SpanContext) (string,
 	return "", false
 }
 
+// NewTracerProvider returns a new instance of an OpenTelemetry tracer provider.
+func NewTracerProvider(ctx context.Context, logger log.Logger, conf []byte) (*tracesdk.TracerProvider, error) {
+	config := Config{}
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, err
+	}
+
+	var exporter *otel_jaeger.Exporter
+	var err error
+
+	if config.Endpoint != "" {
+		var jaegerCollectorEndpointOptions []otel_jaeger.CollectorEndpointOption
+		if config.User != "" {
+			jaegerCollectorEndpointOptions = append(jaegerCollectorEndpointOptions, otel_jaeger.WithUsername(config.User))
+		}
+		if config.Password != "" {
+			jaegerCollectorEndpointOptions = append(jaegerCollectorEndpointOptions, otel_jaeger.WithPassword(config.Password))
+		}
+		jaegerCollectorEndpointOptions = append(jaegerCollectorEndpointOptions, otel_jaeger.WithEndpoint(config.Endpoint))
+
+		exporter, err = otel_jaeger.New(otel_jaeger.WithCollectorEndpoint(jaegerCollectorEndpointOptions...))
+		if err != nil {
+			return nil, err
+		}
+	} else if config.AgentHost != "" && config.AgentPort != 0 {
+		var jaegerAgentEndpointOptions []otel_jaeger.AgentEndpointOption
+		jaegerAgentEndpointOptions = append(jaegerAgentEndpointOptions, otel_jaeger.WithAgentHost(config.AgentHost))
+		jaegerAgentEndpointOptions = append(jaegerAgentEndpointOptions, otel_jaeger.WithAgentPort(strconv.Itoa(config.AgentPort)))
+
+		exporter, err = otel_jaeger.New(otel_jaeger.WithAgentEndpoint(jaegerAgentEndpointOptions...))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		exporter, err = otel_jaeger.New(nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	processor := tracesdk.NewBatchSpanProcessor(exporter)
+	tp := newTraceProvider(ctx, logger, processor, config.SamplerParam, config.ServiceName)
+
+	return tp, nil
+}
+
+func newTraceProvider(ctx context.Context, logger log.Logger, processor tracesdk.SpanProcessor,
```

Comment:
This is what I guess will be a bit more complicate, but I think we will have to go over _all_ configuration options from the existing `jaeger.Config` struct and correlate them with the OTEL exporter configuration, to ensure a compatibility between old and new configs (to the max extent we can). So this will be quite a bit more complex than the Google Cloud tracer.

Some options map clearly (e.g. agent host-port, collector endpoint etc.), but some options need a bit more work. The current  version of the code, for example, assumes we always use ratio-based sampler (and so we're always creating `tracesdk.TraceIDRatioBased`), but we should first inspect `SamplerType` and choose the appropriate sampler type based on that.

Then again, we might find not all options map to the new OTEL config (for example, IIRC, the `RPCMetrics` does not have an equivalent in the OTEL config).

I believe most of the core options can be mapped to the new config, but even with the remaining ones we should be careful and if we cannot find an option equivalent in the OTEL config, we need to state this in the changelog and warn users.

But for now I'd just focus on mapping the ones we can (looks like `Tags`, `Sampler*` and `Reporter*` _could_ be possible).

Reply:
Thanks, I have attempted to factor in some more options from `config_yaml.go`. I _think_ some options such as `SamplerManagerHostPort` and `ReporterFlushInterval` don't have an equivalent in the OTel Go SDK. I'll keep looking, just in case there is an equivalent for these options.

Reply:
Regarding the config options in `jaeger/config_yaml.go`, the config struct has become rather large and unwieldy with all the new options added. I was considering creating new structs for `AgentOptions` and `CollectorOptions` and referencing those in the `Config` struct for a more structured config YAML and readable code. But since existing users might have their configs structured the current way, it might involve some re-work for them.
What do you think?